File: 06_Searching/src/Maximum_Value_At_A_Given_Index_In_A_Bounded_Array_1802.js
/**
 * @param {number} n
 * @param {number} index
 * @param {number} maxSum
 * @return {number}
 */
var maxValue = function(n, index, maxSum) {
    let start = 1;
    let end = maxSum;
    let ans = 0;

    while (start <= end) {
        let mid = Math.floor(start + (end-start)/2);
        let totalSum = mid + sum(index,mid) + sum(n-1-index,mid);

        if (totalSum <= maxSum) {
            ans = mid;
            start = mid + 1;
        } else {
            end = mid - 1;
        }
    }

    return ans;
};

function sum (count,val) {
    let x = Math.min(count,val-1);
    let sum = (val*x) - (x*(x+1))/2;

    sum += Math.max(0,count-val+1);

    return sum;
}

console.log(maxValue(4,2,6))
console.log(maxValue(6,1,10))
console.log(maxValue(8,7,14))
console.log(maxValue(4,0,4))
console.log(maxValue(4,0,395252961))
console.log(maxValue(9,5,24))
console.log(maxValue(9,5,24))

File: 06_Searching/src/Maximum_Value_At_A_Given_Index_In_A_Bounded_Array_1802.java
public class Maximum_Value_At_A_Given_Index_In_A_Bounded_Array_1802 {
    public static int maxValue(int n, int index, int maxSum) {
        long start = 1;
        long end = (long) maxSum;

        long ans = 0;

        while (start <= end) {
            long mid = start + (end-start)/2;
            long totalSum = sum(index,mid) + sum(n-1-index,mid) + mid;

            if (totalSum <= maxSum) {
                ans = mid;
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }

        return (int) ans;
    }
    public static long sum (long count, long val) {
        long x = Math.min(count,val-1);
        long sum = (x*val) - (x * (x+1)) / 2;

        sum += Math.max(0, (count - (val-1)));

        return sum;
    }
}
